sklearn/utils/_param_validation.py

```diff
@@ -0,0 +1,391 @@
+            ]
+
+            # combine the signature and the actual args and kwargs to build a map
+            # param_name: param_value
+            params = {}
+
+            # First, the positional arguments for which we need to parse the signature
+            # to recover their names. The signature objects holds the parameters in the
+            # same order as in the function definition
+            for val, param in zip(args, sig_params):
+                # ignore self for methods
+                if param.name == "self":
+                    continue
+
+                params[param.name] = val
+
+            # Then, the keyword arguments which already hold both the name and the value
+            for name, val in kwargs.items():
+                params[name] = val
+
+            # Finally, the parameters with have a default that are unset
+            for param in [p for p in sig_params if p.default is not p.empty]:
+                if param.name not in params:
+                    params[param.name] = param.default
+
+            validate_parameter_constraints(parameter_constraints, params)
+            return func(*args, **kwargs)
+
+        return wrapper
+
+    return decorator
+
+
+class _Constraint(ABC):
+    """Base class for the constraint objects."""
+
+    @abstractmethod
+    def is_satisfied_by(self, val):
+        """Whether or not a value satisfies the constraint.
+
+        Parameters
+        ----------
+        val : object
+            The value to check.
+
+        Returns
+        -------
+        is_satisfied : bool
+            Whether or not the constraint is satisfied by this value.
+        """
+
+    @abstractmethod
+    def __repr__(self):
+        """A human readable representational string of the constraint."""
+
+
+class StrOptions(_Constraint):
+    """Constraint representing a set of strings.
+
+    Parameters
+    ----------
+    options : set of str
+        The set of valid strings.
+
+    deprecated : set of str or None, default=None
+        A subset of the `options` to mark as deprecated in the repr of the constraint.
+    """
+
+    def __init__(self, options, deprecated=None):
+        self.options = options
+        self.deprecated = deprecated or {}
+
+    def generate_invalid_param_val(self):
+        """Return a value that does not satisfy the constraint."""
+        return "this_is_obviously_an_invalid_val"
+
+    def is_satisfied_by(self, val):
+        return isinstance(val, str) and val in self.options
+
+    def _mark_if_deprecated(self, option):
+        """Add a deprecated mark to an option if needed."""
+        option_str = f"{option!r}"
+        if option in self.deprecated:
+            option_str = f"{option_str} (deprecated)"
+        return option_str
+
+    def __repr__(self):
+        options_str = (
+            f"{', '.join([self._mark_if_deprecated(o) for o in self.options])}"
+        )
+        return f"a str among {{{options_str}}}"
+
+
+class Interval(_Constraint):
```

Comment:
Perhaps, we can use or refactor the Interval class of https://github.com/scikit-learn/scikit-learn/blob/98a9ae05ee5b0663f0c4578786bd9a0be6499a9e/sklearn/_loss/link.py#L15-L20

Reply:
I think that it would be better to separate the object representing a mathematical Interval from constraint object. In that case it would definitely make sense to not keep multiple implementations of the same object